docs/content/latest/develop/query-1-performance/explain-analyze.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Optimizing Queries with EXPLAIN and EXPLAIN ANALYZE
+linkTitle: Optimizing Queries with EXPLAIN and EXPLAIN ANALYZE
+description: Query optimization with EXPLAIN and EXPLAIN ANALYZE
+aliases:
+  - /develop/query-1-performance/explain-analyze/
+headerTitle: Optimizing Queries with EXPLAIN and EXPLAIN ANALYZE
+image: /images/section_icons/index/develop.png
+menu:
+  latest:
+    identifier: explain-analyze
+    parent: query-1-performance
+    weight: 566
+isTocNested: true
+showAsideToc: true
+---
+
+This section describes how to optimize queries using YSQL's `EXPLAIN` and `EXPLAIN ANALYZE` statements.
+
+## The EXPLAIN Statement
+
+For each query that YSQL receives, it creates an execution plan which you can access using the `EXPLAIN` statement. This statement returns two numbers: the so-called startup cost that represents the estimated query cost before the first row is returned, and the total cost to run the query to completion.
+
+You can use the `EXPLAIN` statement in conjunction with `SELECT`, `DELETE`, `INSERT`, `REPLACE`, and `UPDATE` statements. The `EXPLAIN` statement has the following syntax:
+
+```sql
+EXPLAIN [ ( option [, ...] ) ] sql_statement;
+```
+
+The *option* and its values are described in the following table. The most important option is `ANALYZE`.
+
+| Option  | Value                           | Description                                                  |
+| ------- | ------------------------------- | ------------------------------------------------------------ |
+| ANALYZE | boolean                         | Executes the *sql_statement* before returning any run-time statistics.<br/>The output of the *sql_statement* is discarded.<br/>To perform analysis of any data-modifying statement (such as INSERT, UPDATE, and DELETE) without affecting the data, you must wrap EXPLAIN ANALYZE in a transaction using the following syntax:<br/>BEGIN;<br/>EXPLAIN ANALYZE *sql_statement*;<br/>ROLLBACK; |
```

Comment:
The description for ANALYZE is failing to capture a few important things that the postgres documentation describes.

"The ANALYZE option causes the statement to be actually executed, not only planned. The total elapsed time expended within each plan node (in milliseconds) and total number of rows it actually returned are added to the display. This is useful for seeing whether the planner's estimates are close to reality."

One is to make the point that this option, in addition to EXPLAINing the plan used by the statement, it actually executes the statement. And the second is that the reason it is actually executing the statement is collect useful information at the level of each plan node -- like time for the execution and number of rows processed etc.

Reply:
I'm confused: (1) the doc clearly states that ANALYZE "Executes the *sql_statement* ..." and (2) the Examples section of the doc states "In addition to the cost estimates from the query planner, `EXPLAIN ANALYZE` displays the server output produced during the statement execution, as shown in the following example:

```sql
yugabyte=# EXPLAIN ANALYZE SELECT * FROM employees a LEFT JOIN LATERAL (SELECT * FROM employees b WHERE a.a = b.a) c ON TRUE;
                                 QUERY PLAN
-------------------------------------------------------------------------
 Hash Left Join (cost=112.50..390.00 rows=5000 width=8) (actual time=3.939..6.195 rows=100 loops=1)
   Hash Cond: (a.a = b.a)
   ->  Seq Scan on employees a (cost=0.00..100.00 rows=1000 width=4) (actual time=0.568..2.798 rows=100 loops=1)
   ->  Hash  (cost=100.00..100.00 rows=1000 width=4) (actual time=3.363..3.363 rows=100 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 12kB
         ->  Seq Scan on employees b (cost=0.00..100.00 rows=1000 width=4) (actual time=0.458..3.339 rows=100 loops=1)
 Planning Time: 0.939 ms
 Execution Time: 7.089 ms
(8 rows)
```

The server output from the preceding example includes the number of rescans (loops) each node performed, the number of milliseconds passed before the first row was returned, total time before the last tuple was returned by each execution node, and the number of tuples returned by each execution node.

`EXPLAIN`, on the other hand,  does not provide this additional information, as shown in the following examples:

```sql
yugabyte=# EXPLAIN SELECT * FROM employees a LEFT JOIN LATERAL (SELECT * FROM employees b WHERE a.a = b.a) c ON TRUE;
                              QUERY PLAN
----------------------------------------------------------------------
 Hash Left Join (cost=112.50..390.00 rows=5000 width=8)
   Hash Cond: (a.a = b.a)
   ->  Seq Scan on employees a (cost=0.00..100.00 rows=1000 width=4)
   ->  Hash (cost=100.00..100.00 rows=1000 width=4)
         ->  Seq Scan on employees b (cost=0.00..100.00 rows=1000 width=4)
(5 rows)
```

Reply:
Our doc is not structured in the same way Postgres doc is nor contains the same level of detail on purpose, but it seems to capture the essence. Perhaps someone else should provide an opinion. @rkarthik007, would you be able to  review the document, please? It would be very helpful. Thanks!

Reply:
Update the first sentence from:
Executes the *sql_statement* before returning any run-time statistics. The output of the *sql_statement* is discarded.
--> to something like:
The ANALYZE option returns additional useful run-time statistics such as the time spent within each plan node, number of rows processed, etc. It does so by actually executing the statement (not just creating the plan), but the output of *sql_statement* is discarded.